compiler/noirc_frontend/src/elaborator/types.rs

```diff
@@ -76,14 +77,13 @@ impl<'context> Elaborator<'context> {
             FieldElement => Type::FieldElement,
             Array(size, elem) => {
                 let elem = Box::new(self.resolve_type_inner(*elem, kind));
-                let mut size = self.convert_expression_type(size);
-                // TODO(https://github.com/noir-lang/noir/issues/5156): Remove this once we only have explicit numeric generics
-                if let Type::NamedGeneric(type_var, name, _) = size {
-                    size = Type::NamedGeneric(
-                        type_var,
-                        name,
-                        Kind::Numeric(Box::new(Type::default_int_type())),
-                    );
+                let size = self.convert_expression_type(size);
+                if let Type::NamedGeneric(ref _type_var, ref name, ref kind) = size {
+                    if !kind.is_numeric() {
+                        let ident = Ident::new(name.to_string(), span);
+                        self.push_err(ResolverError::UseExplicitNumericGeneric { ident });
+                        return Type::Error;
+                    }
```

Comment:
convert_expression_type should check this so that we don't have to do it at every call site